Add staticcheck to verify-golangci

hlee95 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
1. Adds staticcheck to hack/verify-golangci-lint.sh
2. Fixes or ignores staticcheck failures
3. Removes hack/verify-staticcheck.sh and cleans up hack/tools/go.mod

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/103721

#### Special notes for your reviewer:
There were a lot of failures for SA5011 "possible nil pointer dereference".  The original issue showed 3 occurrences, but after I fixed those with `//nolint:staticcheck`, I found that more showed up the next time I ran the lint check. All of the occurrences of this failure happened on something of the following form:
```
		rs, err := rc.clientSet.AppsV1().ReplicaSets(rc.nsName).Get(context.TODO(), rc.name, metav1.GetOptions{})
		framework.ExpectNoError(err)
		if rs == nil {
			framework.Failf(rsIsNil)
		}
		return int(rs.Status.ReadyReplicas)
```

The linter says that:
```
test/e2e/framework/autoscaling/autoscaling_utils.go:340:17: SA5011: possible nil pointer dereference (staticcheck)
		return int(rs.Status.ReadyReplicas)
		              ^
test/e2e/framework/autoscaling/autoscaling_utils.go:337:6: SA5011(related information): this check suggests that the pointer can be nil (staticcheck)
		if rs == nil {
		   ^
```

I read [here](https://staticcheck.io/docs/checks#SA5011) that the linter is unable to detect that our `framework.Failf()` function results in a panic further down the call chain, which causes the linter to flag all of these cases as a violation of SA5011. The proposed solution from that link is to add a `panic("unreachable")` to `framework.Failf()` so that the linter will understand that a panic occurs, so I did that and it fixed all of the errors except for two. Those two used `Skipf` instead of `Failf`, so I added the `panic("unreachable")` to `Skipf` as well. The other option I thought of was to add `-e SA5011` in hack/verify-golangci-lint.sh but that seemed less ideal because it could hide real errors.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

hlee95 committed:
Fix integer division

hlee95 committed:
Ignore deprecation warnings with //nolint:staticcheck

hlee95 committed:
Update IPVar and IPPortVar functions to have pointer receivers to fix 'ineffective assignment'

hlee95 committed:
Update cpuset.Builder to use pointer receiver

hlee95 committed:
Use time.NewTicker instead of time.Tick to avoid leaking

hlee95 committed:
Add staticcheck to verify-golangci-lint.sh

hlee95 committed:
Remove verify-staticcheck.sh and clean up tools/go.mod

hlee95 committed:
Fix SA5011 errors by making control flow abort obvious to linter

hlee95 commented:
/assign nikhita

dims commented:
/ok-to-test

dims commented:
```
# k8s.io/kubernetes/cmd/kubelet/app/options
cmd/kubelet/app/options/options.go:472:23: cannot use "k8s.io/kubernetes/pkg/util/flag".IPVar{...} (type "k8s.io/kubernetes/pkg/util/flag".IPVar) as type pflag.Value in argument to fs.Var:
```

hlee95 committed:
Convert one more utilflag.IPVar to &utilflag.IPVar

thockin commented on pkg/kubelet/cm/cpuset/cpuset.go:
```diff
@@ -46,7 +46,7 @@ func NewBuilder() Builder {
 // Add adds the supplied elements to the result. Calling Add after calling
 // Result has no effect.
-func (b Builder) Add(elems ...int) {
+func (b *Builder) Add(elems ...int) {
```
This suggests a lack of tests, doesn't it?  Can we tag owners on these when we find them,

nikhita replied:
cc @adisky -- for sig-node related tests

liggitt replied:
yeah... I think `Add()` would actually do the right thing on a struct receiver, but `Result()` would not

hlee95 replied:
Yeah, good call about testing. There does appear to be a test [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/cpuset/cpuset_test.go#L24) but I guess it is not comprehensive enough.

thockin reviewed (commented):
Overall LGTM

hlee95 committed:
Fix last remaining SA5011 error by removing unnecessary r != nil check

MadhavJivrajani commented:
/priority backlog
/label tide/merge-method-squash

dims commented:
/retest

dims commented:
/approve
/assign @liggitt @thockin

liggitt commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -803,10 +803,9 @@ func getEtcdVersionResponse(client *http.Client, url string, target interface{})
 				loopCount--
 				return false, err
 			}
-			//lint:ignore SA5011 If err != nil we are already returning.
```
glad to see this can get dropped... curious if that's because the new static check is more accurate or for some other reason

hlee95 replied:
The staticcheck [docs](https://staticcheck.io/docs/checks#SA5011) say that SA5011 is `A pointer is being dereferenced unconditionally, while also being checked against nil in another place.` So in this particular case, I believe that removing the check for `r != nil` at line 809 makes it so that r is _not_ "also being checked against nil in another place" so SA5011 no longer cares about it.

liggitt commented on cmd/kube-proxy/app/server.go:
```diff
@@ -613,6 +613,7 @@ func serveMetrics(bindAddress, proxyMode string, enableProfiling bool, errCh cha
 	})
 
 	//lint:ignore SA1019 See the Metrics Stability Migration KEP
```
does this mean all the `lint:ignore SA...` comments are no longer paid attention to?

hlee95 replied:
I'm not sure but there is an [issue](https://github.com/golangci/golangci-lint/issues/741) on golangci-lint to that effect. However, I don't see `staticcheck` failures when I run the hack/verify-golangci-lint.sh script on my machine so that seems to suggest that these `//lint:ignore` comments are still being respected or that staticcheck is better than the previous linter and is no longer parking these as errors?

hlee95 replied:
Actually, I think I misunderstood what you were saying. You are right that the new staticcheck is not looking at the old `//lint:ignore`, because without the new `//nolint:staticcheck` comment, staticcheck is failing. I will consider updating the PR to remove the now defunct `//lint:ignore` comments and copying their explanations over to the new `//nolint:staticcheck` comments.

liggitt commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -156,8 +156,7 @@ func (m *manager) Start() {
 	}
 
 	klog.InfoS("Starting to sync pod status with apiserver")
-	//lint:ignore SA1015 Ticker can link since this is only called once and doesn't handle termination.
```
does this mean SA1015 is not running any more?

hlee95 replied:
I removed the `//lint:ignore` comment because I changed line 160 from `time.Tick`, which leaks (according to the description [here](https://pkg.go.dev/time#Tick)), to `time.NewTicker`, which does not leak. I believe SA1015 is still running though, as it is included in the staticcheck [docs](https://staticcheck.io/docs/checks#SA1015).

neolit123 replied:
both would leak if the ticker is not stopped, but i'm assuming that the status manager and this ticker normally have the lifespan of that of the process.

the example here https://pkg.go.dev/time#NewTicker shows how the object should be managed properly.

hlee95 replied:
Ah yes you're right. I could add a `defer syncTicker.Stop()` inside that goroutine at line 162 but I'm not sure if that's worth it since the for select loop will run forever anyway, and because as you and the previous //lint:ignore comment both said, the ticker is only created once.

neolit123 replied:
process lifespan memory leaks can still show up in software that does leak detection...at least for object code from languages without a GC.

the lack of stop can be a problem, if one day the status_manager code is (for some reason) adapted to a dynamic restart but the stop is never added. maybe our static check tools will catch this problem, maybe not.

liggitt reviewed (commented):
all the changes look reasonable, though `verify: golangci-lint` is failing with no useful output in the verify CI job...

hlee95 commented:
Thanks for the comments everyone! The pull-kubernetes-verify job is still failing. When I run the hack/verify-golangci-lint.sh script locally, I do see some failures but none are staticcheck failures and they had been there before I made any changes. Should I be fixing these as well?
```
hannalee@Hannas-MBP kubernetes % hack/verify-golangci-lint.sh
installing golangci-lint 
running golangci-lint 
pkg/generated/openapi/cmd/models-schema/main.go:42:18: GetOpenAPIDefinitions not declared by package openapi (typecheck)
	defs := openapi.GetOpenAPIDefinitions(refFunc)
	                ^
test/integration/framework/controlplane_utils.go:109:65: GetOpenAPIDefinitions not declared by package openapi (typecheck)
	openAPIConfig := genericapiserver.DefaultOpenAPIConfig(openapi.GetOpenAPIDefinitions, openapinamer.NewDefinitionNamer(legacyscheme.Scheme))
	                                                               ^
test/integration/framework/controlplane_utils.go:121:41: GetOpenAPIDefinitions not declared by package openapi (typecheck)
	openAPIConfig.GetDefinitions = openapi.GetOpenAPIDefinitions
	                                       ^
pkg/generated/openapi/openapi_test.go:31:27: undeclared name: `GetOpenAPIDefinitions` (typecheck)
	for name, value := range GetOpenAPIDefinitions(dummyRef) {
	                         ^
test/integration/apiserver/apply/apply_crd_beta_test.go:30:2: "k8s.io/apimachinery/pkg/api/errors" imported but not used as apierrors (typecheck)
	apierrors "k8s.io/apimachinery/pkg/api/errors"
	^
pkg/volume/util/volumepathhandler/volume_path_handler.go:33:2: `losetupPath` is unused (deadcode)
	losetupPath       = "losetup"
	^
pkg/volume/util/hostutil/hostutil.go:78:6: `getFileType` is unused (deadcode)
func getFileType(pathname string) (FileType, error) {
     ^
pkg/util/iptables/iptables_unsupported.go:30:6: `grabIptablesFileLock` is unused (deadcode)
func grabIptablesFileLock(f *os.File) error {
     ^
pkg/volume/util/fsquota/quota.go:48:6: `enabledQuotasForMonitoring` is unused (deadcode)
func enabledQuotasForMonitoring() bool {
     ^
pkg/kubelet/config/file.go:40:2: `podAdd` is unused (deadcode)
	podAdd podEventType = iota
	^
pkg/kubelet/config/file.go:41:2: `podModify` is unused (deadcode)
	podModify
	^
pkg/kubelet/config/file.go:42:2: `podDelete` is unused (deadcode)
	podDelete
	^
cmd/kubelet/app/options/globalflags.go:76:6: `registerDeprecated` is unused (deadcode)
func registerDeprecated(global *flag.FlagSet, local *pflag.FlagSet, globalName, deprecated string) {
     ^
pkg/volume/local/local_test.go:415:6: `testFSGroupMount` is unused (deadcode)
func testFSGroupMount(plug volume.VolumePlugin, pod *v1.Pod, tmpDir string, fsGroup int64) error {
     ^
pkg/volume/awsebs/aws_util.go:204:6: `verifyDevicePath` is unused (deadcode)
func verifyDevicePath(devicePaths []string) (string, error) {
     ^
pkg/volume/awsebs/aws_util.go:219:6: `getDiskByIDPaths` is unused (deadcode)
func getDiskByIDPaths(volumeID aws.KubernetesVolumeID, partition string, devicePath string) []string {
     ^
pkg/kubelet/cm/container_manager.go:197:6: `containerDevicesFromResourceDeviceInstances` is unused (deadcode)
func containerDevicesFromResourceDeviceInstances(devs devicemanager.ResourceDeviceInstances) []*podresourcesapi.ContainerDevices {
     ^
pkg/kubelet/cm/helpers.go:25:6: `hardEvictionReservation` is unused (deadcode)
func hardEvictionReservation(thresholds []evictionapi.Threshold, capacity v1.ResourceList) v1.ResourceList {
     ^
pkg/kubelet/cm/helpers_unsupported.go:54:6: `getCgroupProcs` is unused (deadcode)
func getCgroupProcs(dir string) ([]int, error) {
     ^
pkg/kubelet/kuberuntime/helpers_unsupported.go:22:6: `milliCPUToShares` is unused (deadcode)
func milliCPUToShares(milliCPU int64) int64 {
     ^
pkg/kubelet/dockershim/docker_service.go:64:2: `dockerNetNSFmt` is unused (deadcode)
	dockerNetNSFmt = "/proc/%v/ns/net"
	^
pkg/kubelet/dockershim/helpers.go:57:2: `defaultSeccompOpt` is unused (deadcode)
	defaultSeccompOpt = []dockerOpt{{"seccomp", v1.SeccompProfileNameUnconfined, ""}}
	^
pkg/kubelet/dockershim/security_context.go:63:6: `applyContainerSecurityContext` is unused (deadcode)
func applyContainerSecurityContext(lc *runtimeapi.LinuxContainerConfig, podSandboxID string, config *dockercontainer.Config, hc *dockercontainer.HostConfig, separator rune) error {
     ^
pkg/kubelet/config/file.go:48:2: field `fileName` is unused (unused)
	fileName  string
	^
pkg/kubelet/config/file.go:37:6: type `podEventType` is unused (unused)
type podEventType int
     ^
pkg/kubelet/config/file.go:49:2: field `eventType` is unused (unused)
	eventType podEventType
	^
cmd/kubelet/app/options/globalflags.go:55:6: func `register` is unused (unused)
func register(global *flag.FlagSet, local *pflag.FlagSet, globalName string) {
     ^
pkg/volume/awsebs/aws_util.go:266:6: func `findNvmeVolume` is unused (unused)
func findNvmeVolume(findName string) (device string, err error) {
     ^
pkg/kubelet/kuberuntime/security_context.go:121:6: func `convertToRuntimeCapabilities` is unused (unused)
func convertToRuntimeCapabilities(opts *v1.Capabilities) *runtimeapi.Capability {
     ^
pkg/kubelet/kuberuntime/security_context.go:27:37: func `(*kubeGenericRuntimeManager).determineEffectiveSecurityContext` is unused (unused)
func (m *kubeGenericRuntimeManager) determineEffectiveSecurityContext(pod *v1.Pod, container *v1.Container, uid *int64, username string) *runtimeapi.LinuxContainerSecurityContext {
                                    ^
pkg/kubelet/kuberuntime/security_context.go:107:6: func `convertToRuntimeSELinuxOption` is unused (unused)
func convertToRuntimeSELinuxOption(opts *v1.SELinuxOptions) *runtimeapi.SELinuxOption {
     ^
pkg/kubelet/kuberuntime/security_context.go:81:6: func `convertToRuntimeSecurityContext` is unused (unused)
func convertToRuntimeSecurityContext(securityContext *v1.SecurityContext) *runtimeapi.LinuxContainerSecurityContext {
     ^
pkg/kubelet/kubelet_volumes_test.go:561:27: func `(*stubBlockVolume).GetGlobalMapPath` is unused (unused)
func (f *stubBlockVolume) GetGlobalMapPath(spec *volume.Spec) (string, error) {
                          ^
pkg/kubelet/kubelet_volumes_test.go:565:27: func `(*stubBlockVolume).GetPodDeviceMapPath` is unused (unused)
func (f *stubBlockVolume) GetPodDeviceMapPath() (string, string) {
                          ^
pkg/kubelet/kubelet_volumes_test.go:589:27: func `(*stubBlockVolume).GetMetrics` is unused (unused)
func (f *stubBlockVolume) GetMetrics() (*volume.Metrics, error) {
                          ^
pkg/kubelet/kubelet_volumes_test.go:581:27: func `(*stubBlockVolume).UnmapPodDevice` is unused (unused)
func (f *stubBlockVolume) UnmapPodDevice() error {
                          ^
pkg/kubelet/kubelet_volumes_test.go:573:26: func `stubBlockVolume.MapPodDevice` is unused (unused)
func (f stubBlockVolume) MapPodDevice() error {
                         ^
pkg/kubelet/kubelet_volumes_test.go:577:27: func `(*stubBlockVolume).TearDownDevice` is unused (unused)
func (f *stubBlockVolume) TearDownDevice(mapPath string, devicePath string) error {
                          ^
pkg/kubelet/kubelet_volumes_test.go:556:6: type `stubBlockVolume` is unused (unused)
type stubBlockVolume struct {
     ^
pkg/kubelet/kubelet_volumes_test.go:569:27: func `(*stubBlockVolume).SetUpDevice` is unused (unused)
func (f *stubBlockVolume) SetUpDevice() (string, error) {
                          ^
pkg/kubelet/kubelet_volumes_test.go:585:27: func `(*stubBlockVolume).SupportsMetrics` is unused (unused)
func (f *stubBlockVolume) SupportsMetrics() bool {
                          ^
```

alculquicondor commented:
> defs := openapi.GetOpenAPIDefinitions

It looks like you are missing the generation of some files. Run `make update` once

hlee95 commented:
> > defs := openapi.GetOpenAPIDefinitions
> 
> It looks like you are missing the generation of some files. Run `make update` once

Thanks! I ran and got this:
```
hannalee@Hannas-MBP kubernetes % make update
+++ [0716 09:20:27] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
Generating prerelease lifecycle code for       27 targets
+++ [0716 09:20:38] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
Generating deepcopy code for      234 targets
+++ [0716 09:20:44] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
Generating defaulter code for       93 targets
+++ [0716 09:20:52] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
Generating conversion code for      128 targets
+++ [0716 09:21:09] Building go targets for darwin/amd64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
Generating openapi code for KUBE
Generating openapi code for AGGREGATOR
Generating openapi code for APIEXTENSIONS
Generating openapi code for CODEGEN
Generating openapi code for SAMPLEAPISERVER
Running in silent mode, run with SILENT=false if you want to see script logs.
Running in short-circuit mode; run with FORCE_ALL=true to force all scripts to run.
Running update-generated-protobuf
Running update-generated-protobuf FAILED
```

It failed on `update-generated-protobuf`. I am probably missing some developer setup steps as this is my first contribution. Can you point me to what I should do to fix?

alculquicondor commented:
Not sure. You didn't get any specific error message?

hlee95 commented:
> Not sure. You didn't get any specific error message?

I ran it again with SILENT=false and it said:
```
Running update-generated-protobuf
+++ [0716 09:51:03] Verifying Prerequisites....
  !!! Cannot find GNU tar. Build on Linux or install GNU tar
      on Mac OS X (brew install gnu-tar).
Running update-generated-protobuf FAILED
make: *** [update] Error 1
```

I am on a new machine haha so I didn't have anything installed. I ended up needing to do some basic things like:
* `brew install coreutils ed findutils gawk gnu-sed gnu-tar grep make` from the developer setup steps [here](https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#setting-up-macos)
* Install Docker
* `brew install jq`
* ./hack/install-etcd.sh and 

and now I am seeing an error that I don't know how to address:
```
hannalee@Hannas-MBP kubernetes % make update SILENT=false
Running in short-circuit mode; run with FORCE_ALL=true to force all scripts to run.
Running update-generated-protobuf
...
Running update-codegen
...
Running update-generated-runtime
...
Running update-generated-device-plugin
...
Running update-generated-api-compatibility-data
...
Running update-generated-docs
Generating prerelease lifecycle code for       27 targets
Generating deepcopy code for      193 targets
Generating defaulter code for       88 targets
Generating conversion code for      115 targets
Generating openapi code for KUBE
Generating openapi code for AGGREGATOR
Generating openapi code for APIEXTENSIONS
+++ [0716 12:05:19] Building go targets for darwin/amd64:
    cmd/gendocs
    cmd/genkubedocs
    cmd/genman
    cmd/genyaml
/Users/hannalee/go/src/k8s.io/kubernetes/kubernetes/hack/lib/golang.sh: line 655: 79729 Illegal instruction: 4  go install "${build_args[@]}" "$@"
!!! [0716 12:05:25] Call tree:
!!! [0716 12:05:25]  1: /Users/hannalee/go/src/k8s.io/kubernetes/kubernetes/hack/lib/golang.sh:731 kube::golang::build_some_binaries(...)
!!! [0716 12:05:25]  2: /Users/hannalee/go/src/k8s.io/kubernetes/kubernetes/hack/lib/golang.sh:875 kube::golang::build_binaries_for_platform(...)
!!! [0716 12:05:25]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0716 12:05:25] Call tree:
!!! [0716 12:05:25]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0716 12:05:25] Call tree:
!!! [0716 12:05:25]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [all] Error 1
Running update-generated-docs FAILED
make: *** [update] Error 1
```

hlee95 committed:
Add more //nolint:staticcheck for failures caught in PR tests

hlee95 commented:
I expanded the full logs of the test failure and found that there were still a few more staticcheck failures (see below). I'm not sure why these didn't show up when I ran the hack/verify-golangci-lint.sh script locally, but either way I pushed a new commit that hopefully will fix it.

```
running golangci-lint 
test/integration/garbagecollector/garbage_collector_test.go:602:7: SA1019: options.OrphanDependents is deprecated: please use the PropagationPolicy, this field will be deprecated in 1.7. Should the dependent objects be orphaned. If true/false, the "orphan" finalizer will be added to/removed from the object's finalizers list. Either this field or PropagationPolicy may be set, but not both. +optional (staticcheck)
	case options.OrphanDependents == nil && options.PropagationPolicy == nil && len(initialFinalizers) == 0:
	     ^
test/integration/garbagecollector/garbage_collector_test.go:605:7: SA1019: options.OrphanDependents is deprecated: please use the PropagationPolicy, this field will be deprecated in 1.7. Should the dependent objects be orphaned. If true/false, the "orphan" finalizer will be added to/removed from the object's finalizers list. Either this field or PropagationPolicy may be set, but not both. +optional (staticcheck)
	case options.OrphanDependents != nil:
	     ^
test/integration/garbagecollector/garbage_collector_test.go:607:13: SA1019: options.OrphanDependents is deprecated: please use the PropagationPolicy, this field will be deprecated in 1.7. Should the dependent objects be orphaned. If true/false, the "orphan" finalizer will be added to/removed from the object's finalizers list. Either this field or PropagationPolicy may be set, but not both. +optional (staticcheck)
		orphan = *options.OrphanDependents
		          ^
test/integration/scheduler_perf/scheduler_perf_legacy_test.go:443:3: SA3001: should not assign to b.N (staticcheck)
		b.N = minPods
		^
pkg/volume/util/fsquota/quota_linux.go:307:50: SA4009: argument poduid is overwritten before first use (staticcheck)
func AssignQuota(m mount.Interface, path string, poduid types.UID, bytes *resource.Quantity) error {
                                                 ^
pkg/volume/util/fsquota/quota_linux.go:322:2: SA4009(related information): assignment to poduid (staticcheck)
	poduid = types.UID(uuid.NewUUID())
	^
+++ exit code: 1
+++ error: 1
FAILED   verify-golangci-lint.sh	428s
```

nikhita commented:
/retest

nikhita commented:
/retest

neolit123 commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -803,10 +803,9 @@ func getEtcdVersionResponse(client *http.Client, url string, target interface{})
 			}
-			//lint:ignore SA5011 If err != nil we are already returning.
 			defer r.Body.Close()
 
-			if r != nil && r.StatusCode >= 500 && r.StatusCode <= 599 {
+			if r.StatusCode >= 500 && r.StatusCode <= 599 {
```
LGTM on the /cmd/kubeadm change.

neolit123 commented on test/e2e/apimachinery/garbage_collector.go:
```diff
@@ -67,7 +67,7 @@ func estimateMaximumPods(c clientset.Interface, min, max int32) int32 {
 	}
 	//avoid creating exactly max pods
-	availablePods *= 8 / 10
+	availablePods = int32(float32(availablePods) * 8.0 / 10.0)
```
does this function have unit tests, because this is a change in behavior?
the original code has a problem; the rvalue `8 / 10` is always zero as per the lvalue type of int32.

```
	var availablePods int32
	availablePods = 15
	availablePods *= 8 / 10
	fmt.Println(availablePods) // 0
	availablePods = 15
	availablePods = int32(float32(availablePods) * 8.0 / 10.0)	
	fmt.Println(availablePods) // 12
```

also the correct order of operations in your change should not be `*` then `/`.
(most) compilers evaluate the rvalue before the operator+assignment with the lvalue.

it should be `availablePods = int32(float32(availablePods) * (8.0 / 10.0))`.
```
	var availablePods float32
	availablePods = 15
	availablePods *= 10.0
	availablePods /= 8.0
	fmt.Println(availablePods) // 18.75
	availablePods = 15
	availablePods *= 8.0 / 10.0
	fmt.Println(availablePods) // 12
```

but since the original code is buggy it's unclear whether truncating the float is the desired result.
possibly, yes.

hlee95 replied:
I don't see any unit tests for this function; it itself is part of a test. And yes, it was always being set to zero before (and then would be set to `min` at line 76), which I too assumed was a bug and I think the desired behavior was to multiply it by .8 but we could check with the original author if we want to be sure.

Thanks for the catch on the order of operations - I will update that!

neolit123 replied:
it can be changed to ` * 0.8` too.
original commit is here:
https://github.com/kubernetes/kubernetes/commit/91cc7b94c8f4af77c5fd8814761419de2fb63ac8

hlee95 replied:
@smarterclayton the commit in question is from 4 years ago haha but any chance you could give a confirmation as to whether it's ok here to change the behavior to actually multiply by 0.8 instead of setting to 0? Maybe silly to ask but just want to be sure :)

neolit123 commented on test/e2e/framework/log.go:
```diff
@@ -49,6 +49,7 @@ func Failf(format string, args ...interface{}) {
 	skip := 2
 	log("FAIL", "%s\n\nFull Stack Trace\n%s", msg, PrunedStack(skip))
 	e2eginkgowrapper.Fail(nowStamp()+": "+msg, skip)
+	panic("unreachable")
```
is it worth using static check pragmas here instead of panicing?
e.g. ` //nolint:staticcheck`

hlee95 replied:
I did start by adding the `//nolint:staticcheck` in several places but there ended up being many occurrences of SA5011 (I kept fixing a few and then more would show up the next time I linted; maybe the output was truncated) so I decided that adding the panic would be simpler. Adding the `panic("unreachable")` also appeared to be a recommended approach, from the docs that I linked in the PR description up top, for handling a case where staticcheck doesn't understand that the control flow has been aborted.

neolit123 commented on test/integration/scheduler_perf/scheduler_perf_legacy_test.go:
```diff
@@ -440,7 +440,7 @@ func benchmarkScheduling(numExistingPods, minPods int,
 	if b.N < minPods {
 		//lint:ignore SA3001 Set a minimum for b.N to get more meaningful results
-		b.N = minPods
+		b.N = minPods //nolint:staticcheck
```
this is clamping the minimum of the b.N.

can we remove the old //lint:....` line?
why does the new static check need the pragma?

hlee95 replied:
There is an existing issue about staticcheck not respecting `//lint:ignore` when run from `golangci-lint`: https://github.com/golangci/golangci-lint/issues/741 but I don't know more beyond that. Using `//nolint:...` is what's suggested in the golangci-lint [docs](https://golangci-lint.run/usage/false-positives/).

Open to any suggestions on how to proceed. I originally left the old `//lint:ignore` pragmas in because they had more information than my new `//nolint:staticcheck` pragmas. The two things I am thinking of now are either either:
* deleting the `//lint:ignore`s but copying over their extra comments to the new `//nolint:staticcheck`s
* leaving the `//lint:ignore`s there so that if/when the issue I linked is fixed, we can go back to using those and we can delete the extra `//nolint:staticcheck`s in the future

neolit123 replied:
my vote is for the first option, with a comment of why this is being done (error message from the linter).

hlee95 replied:
Ok sounds good. I will update them all later tonight or tomorrow.

hlee95 committed:
Fix order of operations

hlee95 committed:
Simplify multiplication

fedebongio commented:
/remove-sig api-machinery

hlee95 committed:
Remove //lint:ignore pragmas that aren't being used anymore

hlee95 committed:
Add syncTicker.Stop()

hlee95 commented:
/retest

hlee95 commented:
/retest

fedebongio commented:
/remove-sig api-machinery

dims commented:
/ok-to-test

dims commented:
@liggitt this looks ready to land!

liggitt commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -156,10 +156,10 @@ func (m *manager) Start() {
+	syncTicker := time.NewTicker(syncPeriod)
 	// syncPod and syncBatch share the same go routine to avoid sync races.
 	go wait.Forever(func() {
+		defer syncTicker.Stop()
```
calling wait.Forever adds a panic recovery path that can re-call the provided function. If that ever occurs, this change will break the code because it will have stopped the ticker after the function exited the first time

hlee95 replied:
@liggitt  Got it, thanks. Should I move the `syncTicker := time.NewTicker(syncPeriod)` into the wait.Forever function then so that a new ticker is created if the function is re-called? Or what would you recommend here?

liggitt replied:
no, I would keep the lint:ignore directive

aojea commented:
/retest

aojea commented:
/lgtm

thockin commented:
It should work if you run on the vendor/... links to the staging repos?

aojea commented:
> It should work if you run on the vendor/... links to the staging repos?

yeah, I think that is the missing piece

hlee95 commented:
> @hlee95 could you please rebase this to latest master?

Hi yes, sorry for the long delay here. Do you mean rebase my branch onto latest master or merge in the latest master into my branch? I'm afraid if I rebase then I will have to force-push and then the PR discussion will look confusing.

thockin commented:
I don't see a way around that sadly.

On Tue, Nov 2, 2021 at 5:56 AM Hanna Lee ***@***.***> wrote:

> @hlee95 <https://github.com/hlee95> could you please rebase this to
> latest master?
>
> Hi yes, sorry for the long delay here. Do you mean rebase my branch onto
> latest master or merge in the latest master into my branch? I'm afraid if I
> rebase then I will have to force-push and then the PR discussion will look
> confusing.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/103723#issuecomment-957507867>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVDMHWMUBLJGCGX57SLUJ7UY7ANCNFSM5AOH5YUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

hlee95 committed:
Use nolint directive instead of stopping ticker, per liggit's suggestion

aojea commented:
/retest

aojea commented:
LGTM , one question,
should we keep the previous static check running in parallel?

https://github.com/kubernetes/kubernetes/pull/103723/commits/907a3a1066e0ed85fa59ad72b0ff46bc8053afe9

at least until we fix the staging modules problem https://github.com/kubernetes/kubernetes/pull/103723#issuecomment-901189371 ?

@thockin is experimenting with something to solve it https://github.com/kubernetes/kubernetes/pull/105891

thockin commented:
My experiments are very experimental :)

On Thu, Nov 4, 2021 at 1:00 AM Antonio Ojea ***@***.***>
wrote:

> LGTM , one question,
> should we keep the previous static check running in parallel?
>
> 907a3a1
> <https://github.com/kubernetes/kubernetes/commit/907a3a1066e0ed85fa59ad72b0ff46bc8053afe9>
>
> at least until we fix the staging modules problem #103723 (comment)
> <https://github.com/kubernetes/kubernetes/pull/103723#issuecomment-901189371>
> ?
>
> @thockin <https://github.com/thockin> is experimenting with something to
> solve it #105891 <https://github.com/kubernetes/kubernetes/pull/105891>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/103723#issuecomment-960531150>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVDJMRBXXT4FRF3XK4TUKI4TPANCNFSM5AOH5YUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

fedebongio commented:
/remove-sig api-machinery

thockin commented:
Are we all good with this one?

aojea commented:
I suggest to keep both statickcheck in parallel until we solve the staging modules problem for golangci, 

https://github.com/kubernetes/kubernetes/pull/103723#issuecomment-960531150

BenTheElder commented:
I haven't had a chance to look into that problem, can we just update the script to run golangci-lint once per module? Or it will break because of our symlink hackery?

thockin commented:
I have a branch where I am trying to make ruleguard work (part of
golangci-line).  It includes the following diff.   It moves list of linters
to a config file, but you can see past that here:

```
diff --git a/hack/verify-golangci-lint.sh b/hack/verify-golangci-lint.sh
index 1e0bc1192ed..75b226dace0 100755
--- a/hack/verify-golangci-lint.sh
+++ b/hack/verify-golangci-lint.sh
@@ -38,16 +38,25 @@ export GO111MODULE=on
 # Install golangci-lint
 echo 'installing golangci-lint '
 pushd "${KUBE_ROOT}/hack/tools" >/dev/null
+  go mod download
   go install github.com/golangci/golangci-lint/cmd/golangci-lint
 popd >/dev/null

 cd "${KUBE_ROOT}"

+# This is a bug in golangci-lint wrt ruleguard
+# https://github.com/golangci/golangci-lint/issues/1999
+golangci-lint cache clean
+
+# The config is in ${KUBE_ROOT}/.golangci.yaml
 echo 'running golangci-lint '
-golangci-lint run \
-  --timeout 30m \
-  --disable-all \
-  -E deadcode \
-  -E unused \
-  -E varcheck \
-  -E ineffassign
+if [[ "$#" > 0 ]]; then
+    golangci-lint run "$@"
+else
+    golangci-lint run ./...
+    for d in staging/src/k8s.io/*; do
+        pushd ./vendor/k8s.io/$(basename "$d") >/dev/null
+        golangci-lint run ./...
+        popd >/dev/null
+    done
+fi
```

On Tue, Nov 9, 2021 at 11:52 AM Benjamin Elder ***@***.***>
wrote:

> I haven't had a chance to look into that problem, can we just update the
> script to run golangci-lint once per module? Or it will break because of
> our symlink hackery?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/103723#issuecomment-964483797>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVBODJ6TLOVKRRWDQWTULF3XXANCNFSM5AOH5YUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

BenTheElder commented:
> it moves list of linters to a config file,

yeah, there's another PR doing this, I think we should do this (https://github.com/kubernetes/kubernetes/pull/104835)

```
+if [[ "$#" > 0 ]]; then
+    golangci-lint run "$@"
+else
+    golangci-lint run ./...
+    for d in staging/src/k8s.io/*; do
+        pushd ./vendor/k8s.io/$(basename "$d") >/dev/null
+        golangci-lint run ./...
+        popd >/dev/null
+    done
+fi
```

And this is working as intended?
We should do this instead of keeping around the old staticcheck script.

Maybe we should break some of this out into one PR that handles covering staging and moving to a config file (or maybe two PRs?) so we can avoid all of these other PRs adding new linters stomping each other in script changes versus just the config file.

thockin committed:
Move golangci-lint config to a file

Make verify-golangci-lint.sh work across modules and take an optional
argument for a package.

thockin committed:
Move golangci-lint config to a file

Make verify-golangci-lint.sh work across modules and take an optional
argument for a package.

thockin committed:
Add staticcheck to .golangci.yaml

TODO: port options from verify-staticcheck.sh and delete it

hlee95 committed:
Copy in options from now-deleted hack/verify-staticcheck.sh script

hlee95 commented:
> I moved to separate commits - @hlee95 do you want to steal these commits?
> 
> [thockin@6931cba](https://github.com/thockin/kubernetes/commit/6931cbaff4945e2cab28d44acf08b30db31eeb23) [thockin@64a6770](https://github.com/thockin/kubernetes/commit/64a67701771ad69ca3c2bf3186313916b42f0374)

Yup I cherry-picked those.

> The second does NOT move all the options from the staticcheck script or delete it - I assume you can cover that?

I copied over the `checks` and `ignore_pattern` options from the old script; please let me know if I did it correctly and if I missed anything.

hlee95 commented:
/retest

BenTheElder commented:
unrelated to this PR I think but https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103723/pull-kubernetes-verify/1458620020010520576 does not capture useful results in the junit, only the go mod downloads (which ... are honestly noise ...), I think the golangci-lint errors are not winding up in stdout instead of stderr, so to find them you have to expand the whole build log and read through it.

we should fix that

aojea commented:
it's a bit ironic :upside_down_face: 
```
In ./hack/verify-golangci-lint.sh line 48:
if [[ "$#" > 0 ]]; then
           ^-- SC2071: > is for string comparisons. Use -gt instead.


In ./hack/verify-golangci-lint.sh line 53:
        pushd ./vendor/k8s.io/$(basename "$d") >/dev/null
                              ^--------------^ SC2046: Quote this to prevent word splitting.
```

thockin commented on .golangci.yaml:
```diff
@@ -17,6 +17,15 @@ linters:
+      "-ST1*",  # Mostly stylistic, redundant w/ golint
+      "-SA5011" # Possible nil pointer dereference
+    ]
+    ignore_pattern: [
```
Is this a thing?  I don't see it in https://golangci-lint.run/usage/configuration/

hlee95 replied:
It is indeed not a thing, I will remove it.

thockin commented:
@aojea comments fixed in https://github.com/thockin/kubernetes/commit/395bcb8beee41d367399adbab14f5557a140b4af

fedebongio commented:
/triage accepted
/remove-sig api-machinery

thockin commented:
I found another bug in script - testing fix now.  Basically errexit terminates if any of the runs fails, and doesn't run the rest.

thockin commented:
https://github.com/thockin/kubernetes/commit/5758c48c3feec82b125a5dbdc7a4c1a18a1ec28d

aojea commented:
/remove priority-backlog
/priority important-soon

Once this is in we have to spend some time getting this right, is easier to prevent than to find needless in haystacks

aojea commented:
why the linter doesn't catch these "capture variable on loop" things ?
https://github.com/kubernetes/kubernetes/pull/106373

thockin commented:
Code freeze looming - anything we can do to help make this work?

BenTheElder commented:
@aojea https://github.com/kubernetes/kubernetes/pull/104835 will cover those.

aojea commented:
can we justify to get this during the test-freeze period? I think is legit

aojea commented:
trying to get this to the finish line https://github.com/kubernetes/kubernetes/pull/106448
keeping all the author on the commits of course, great job here

hlee95 committed:
Remove invalid option

thockin commented:
https://github.com/kubernetes/kubernetes/pull/106448 supercedes.
